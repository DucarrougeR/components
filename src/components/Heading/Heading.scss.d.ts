export const lensHeading: string;
export const isDisplay1: string;
export const isDisplay2: string;
export const isDisplay3: string;
export const isHeading1: string;
export const isHeading2: string;
export const isHeading3: string;
export const isHeading4: string;
export const isHeading5: string;
export const isHeading6: string;
export const fontWeightLight: string;
export const fontWeightNormal: string;
export const fontWeightSemiBold: string;
export const fontWeightBold: string;
export const fontWeightExtraBold: string;
export const transformUpper: string;
export const transformLower: string;
export const transformCaps: string;
export const transformNone: string;
