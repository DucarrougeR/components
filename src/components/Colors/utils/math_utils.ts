/**
 * Utility to convert a numerical value to a percent given the value and a max value.
 */
export const toPerc = (value: number, maxValue: number) =>
  Math.round((value / maxValue) * 100)
