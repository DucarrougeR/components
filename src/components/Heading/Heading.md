```js
<Heading className="cool" size="d1">Hello</Heading>
<Heading size="d2" level="2">Hello</Heading>
<Heading size="d3" level="3">Hello</Heading>
<Heading size="1" level="1">Hello</Heading>
<Heading size="2" level="2">Hello</Heading>
<Heading size="3" level="3">Hello</Heading>
<Heading size="4" level="4">Hello</Heading>
<Heading size="5" level="5">Hello</Heading>
<Heading level="6" size="">Hello</Heading>
```
