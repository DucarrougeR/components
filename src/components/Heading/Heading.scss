@import '../../styles/util';
@import '../../styles/variables';

.lens-heading {
  margin: 0;
  text-rendering: optimizeLegibility;
  font-family: $brand-font;
  font-weight: $default-font-weight;
}

.is-display-1 {
  font-size: font-size(d1);
  line-height: line-height(d1);
}

.is-display-2 {
  font-size: font-size(d2);
  line-height: line-height(d2);
}

.is-display-3 {
  font-size: font-size(d3);
  line-height: line-height(d3);
}

h1, .is-heading-1 {
  font-size: font-size(1);
  line-height: line-height(1);
}

h2, .is-heading-2 {
  font-size: font-size(2);
  line-height: line-height(2);
}

h3, .is-heading-3 {
  font-size: font-size(3);
  line-height: line-height(3);
}

h4, .is-heading-4 {
  font-size: font-size(4);
  line-height: line-height(4);
}

h5, .is-heading-5 {
  font-size: font-size(5);
  line-height: line-height(5);
}

h6, .is-heading-6 {
  font-size: font-size(6);
  line-height: line-height(6);
}

.font-weight-light {
  font-weight: 300;
}

.font-weight-normal {
  font-weight: 400;
}

.font-weight-semi-bold {
  font-weight: 600;
}

.font-weight-bold {
  font-weight: 700;
}

.font-weight-extra-bold {
  font-weight: 600;
}

.transform-upper {
  text-transform: uppercase;
}

.transform-lower {
  text-transform: lowercase;
}

.transform-caps {
  text-transform: capitalize;
}

.transform-none {
  text-transform: none;
}

