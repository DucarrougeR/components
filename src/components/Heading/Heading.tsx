import * as React from 'react'
import styled, { css } from '../../styled_components'
import { truncate } from '../../styles/typography'
import { lineHeights } from '../../styles/line_heights'
import { fontWeights } from '../../styles/font_weights'
import { RampSizes } from '../../styles/ramp_sizes'

export enum HeadingAlignments {
  Left = 'left',
  Center = 'center',
  Right = 'right'
}

export enum HeadingLevels {
  L1 = '1',
  L2 = '2',
  L3 = '3',
  L4 = '4',
  L5 = '5',
  L6 = '6'
}

export enum HeadingTextTransforms {
  Caps = 'caps',
  Lower = 'lower',
  None = 'none',
  Upper = 'upper'
}

export enum HeadingWeights {
  Bold = 'bold',
  ExtraBold = 'extraBold',
  Light = 'light',
  Normal = 'normal',
  SemiBold = 'semiBold'
}

export interface HeadingGeneratorProps {
  /** Headling tag level mapping for h1-h6 */
  level?: HeadingLevels
}

export interface HeadingProps extends HeadingGeneratorProps {
  /** Size mapping from type ramp */
  size?: RampSizes
  /** Font weight */
  weight?: HeadingWeights
  /** Text tranform  */
  transform?: HeadingTextTransforms
  /** Text align */
  align?: HeadingAlignments
  className?: string
  /** Truncate text on overflow */
  truncate?: boolean
}

const convertHeadingLevelToRampSize = (level: HeadingLevels | undefined) => {
  switch (level) {
    case HeadingLevels.L1:
      return RampSizes.One
    case HeadingLevels.L2:
      return RampSizes.Two
    case HeadingLevels.L3:
      return RampSizes.Three
    case HeadingLevels.L4:
      return RampSizes.Four
    case HeadingLevels.L5:
      return RampSizes.Five
    case HeadingLevels.L6:
      return RampSizes.Six
    default:
      return RampSizes.Three
  }
}

/**
 * Headings are used to help users understand  what a major section of an interface is about, for example the labeling
 * of a page or a title of a card component.
 */
const HeadingGenerator: React.SFC<HeadingProps> = ({ level, ...args }) => {
  // This prevents our props from being passed directly to the underlying h* tags, which ultimately
  // would cause some warnings. Ideally we would define the return type for this function, and
  // Typescript would warn us when passing props that are invalid.
  //
  // See https://reactjs.org/warnings/unknown-prop.html
  const props = Object.assign({}, args)
  delete props.align
  delete props.size
  delete props.transform
  delete props.truncate
  delete props.weight

  switch (level) {
    case HeadingLevels.L1:
      return <h1 {...props}>{props.children}</h1>
    case HeadingLevels.L2:
      return <h2 {...props}>{props.children}</h2>
    case HeadingLevels.L4:
      return <h4 {...props}>{props.children}</h4>
    case HeadingLevels.L5:
      return <h5 {...props}>{props.children}</h5>
    case HeadingLevels.L6:
      return <h6 {...props}>{props.children}</h6>
    case HeadingLevels.L3:
    default:
      return <h3 {...props}>{props.children}</h3>
  }
}

function textTransform(transform: HeadingTextTransforms | undefined) {
  switch (transform) {
    case HeadingTextTransforms.Upper:
      return css`
        text-transform: uppercase;
      `
    case HeadingTextTransforms.Lower:
      return css`
        text-transform: lowercase;
      `
    case HeadingTextTransforms.Caps:
      return css`
        text-transform: capitalize;
      `
    case HeadingTextTransforms.None:
    default:
      return css`
        text-transform: none;
      `
  }
}

function alignment(align: HeadingAlignments | undefined) {
  return css`
    text-align: ${align || HeadingAlignments.Left};
  `
}

export const Heading = styled<HeadingProps>(HeadingGenerator)`
  font-size: ${props =>
    props.theme.fontRamp[
      props.size || convertHeadingLevelToRampSize(props.level)
    ]};
  line-height: ${props =>
    lineHeights[props.size || convertHeadingLevelToRampSize(props.level)]};
  font-weight: ${props => fontWeights[props.weight || HeadingWeights.Normal]};
  ${props => textTransform(props.transform)}
  ${props => alignment(props.align)}
  ${props => truncate(props.truncate || false)}
`
