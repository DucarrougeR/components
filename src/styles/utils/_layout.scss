// Returns a value from the $spacing list
@function spacing($key) {
  @return map-lookup($spacing, $key, '$spacing');
}
