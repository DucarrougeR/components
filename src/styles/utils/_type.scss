// Returns a value from the $test-size list
@function font-size($key) {
  @return map-lookup($text-size, $key, '$text-size');
 }

 // Returns a value from the $line-height list
 @function line-height($key) {
   @return map-lookup($line-height, $key, '$line-height');
 }

 @mixin fontSize($size) {
  font-size: $size;
  font-size: pxToRem($size);
}
