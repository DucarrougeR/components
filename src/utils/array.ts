/**
 * Returns an array of length `size` filled with zeros.
 * Useful for looping constructs where a length is known, but no array is provided.
 */

export const sizedArray = (size: number) => Array(size).fill(0)
