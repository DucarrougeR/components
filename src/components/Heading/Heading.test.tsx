import * as React from 'react'
import { createWithTheme } from '../../../test/utils/create_with_theme'
import { Heading } from './Heading'

test('A default Heading', () => {
  const component = createWithTheme(<Heading>ðŸ¥‘</Heading>)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('A <h1> Heading', () => {
  const component = createWithTheme(<Heading is="h1">ðŸ¥‘</Heading>)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('A <h1> Heading sized to <h2>', () => {
  const component = createWithTheme(
    <Heading is="h1" size="xlarge">
      ðŸ¥‘
    </Heading>
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('A Heading to bold', () => {
  const component = createWithTheme(<Heading weight="bold">ðŸ¥‘</Heading>)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('A Heading transformed', () => {
  const component = createWithTheme(<Heading transform="caps">ðŸ¥‘</Heading>)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('A Heading truncated', () => {
  const component = createWithTheme(<Heading truncate>ðŸ¥‘</Heading>)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
