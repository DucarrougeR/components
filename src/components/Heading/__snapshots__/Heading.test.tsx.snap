// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A <h1> Heading 1`] = `
.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  margin: 0;
  font-size: 1.5625rem;
  line-height: 2.5rem;
  font-weight: 400;
  text-transform: none;
  text-align: left;
}

<h1
  className="c0"
>
  ðŸ¥‘
</h1>
`;

exports[`A <h1> Heading sized to <h2> 1`] = `
.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  margin: 0;
  font-size: 1.375rem;
  line-height: 2rem;
  font-weight: 400;
  text-transform: none;
  text-align: left;
}

<h1
  className="c0"
>
  ðŸ¥‘
</h1>
`;

exports[`A Heading to semi-bold 1`] = `
.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  margin: 0;
  font-size: 1.125rem;
  line-height: 1.75rem;
  font-weight: 700;
  text-transform: none;
  text-align: left;
}

<h3
  className="c0"
>
  ðŸ¥‘
</h3>
`;

exports[`A Heading transformed 1`] = `
.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  margin: 0;
  font-size: 1.125rem;
  line-height: 1.75rem;
  font-weight: 400;
  text-transform: capitalize;
  text-align: left;
}

<h3
  className="c0"
>
  ðŸ¥‘
</h3>
`;

exports[`A default Heading 1`] = `
.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  margin: 0;
  font-size: 1.125rem;
  line-height: 1.75rem;
  font-weight: 400;
  text-transform: none;
  text-align: left;
}

<h3
  className="c0"
>
  ðŸ¥‘
</h3>
`;
